src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -676,38 +676,191 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
+            internal DeadEnumeratorWrapper(IEnumerator<TSource> corpse)
+                : this(corpse, null)
+            {
+            }
+
+            internal DeadEnumeratorWrapper(IEnumerator<TSource> corpse, ExceptionDispatchInfo exception)
+            {
+                _corpse = corpse;
+                _exception = exception;
+            }
+
+            public TSource Current
+            {
+                get { return default(TSource); }
+            }
+
+            object IEnumerator.Current
+            {
+                get { return Current; }
+            }
+
+            public bool MoveNext()
+            {
+                if (_exception != null)
+                {
+                    ExceptionDispatchInfo ex = _exception;
+                    _exception = null;
+                    ex.Throw();
+                }
+                return false;
+            }
+
+            public void Dispose()
+            {
+                _corpse.Dispose();
+            }
+
+            public void Reset()
+            {
+                throw new NotSupportedException();
+            }
+        }
+
         public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> source, int count)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            return SkipIterator<TSource>(source, count);
+            return new SkipIterator<TSource>(source, count);
         }
 
-        private static IEnumerable<TSource> SkipIterator<TSource>(IEnumerable<TSource> source, int count)
+        private sealed class SkipIterator<TSource> : IEnumerable<TSource>
         {
-            using (IEnumerator<TSource> e = source.GetEnumerator())
+            private readonly IEnumerable<TSource> _source;
+            private readonly int _count;
+
+            public SkipIterator(IEnumerable<TSource> source, int count)
             {
-                while (count > 0 && e.MoveNext()) count--;
-                if (count <= 0)
+                _source = source;
+                _count = count > 0 ? count : 0;
+            }
+
+            public IEnumerator<TSource> GetEnumerator()
+            {
+                IEnumerator<TSource> sourceEnumerator = _source.GetEnumerator();
+                try
                 {
-                    while (e.MoveNext()) yield return e.Current;
+                    for (int count = _count; count != 0; --count)
+                    {
+                        // It should be harmless to have the recipient call MoveNext again.
+                        // but we guard against strange behaviour from shortcuts in the enumerator
+                        // It should also be harmless to dispose the enumerator here if we've
+                        // exhausted it, but we guard against strange behaviour in either the
+                        // enumerator or the caller.
+                        if (!sourceEnumerator.MoveNext()) return new DeadEnumeratorWrapper<TSource>(sourceEnumerator);
```

Comment:
The other problem with ExceptionDispatchInfo is debuggability. It preserves stack trace, but not stack itself. Basically if you have to do a postmortem analysis of a crash dump, you will not be able to dig through the application state at the time of the crash. You will have the state preserved at the time of ex.Throw() which is not all that useful. 
ExceptionDispatchInfo is a necessary evil in async. It is still already inconvenient enough to force us using exception filters in Roslyn to guarantee robust and immediate failfast in async code.
In iterators the dead iterator trick with ExceptionDispatchInfo would be just to gain some perf and IMO is probably not worth the overall trouble.